File: leetcode/0001_two_sum.kt
package leetcode

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val result = IntArray(2)

        val seen = mutableMapOf<Int, Int>()
        nums.forEachIndexed { i, num ->
            val j = seen.get(target - num)
            if (j != null) {
                result[0] = j
                result[1] = i
            }

            seen.put(num, i)
        }

        return result
    }
}

fun main(args: Array<String>) {
    val solution = Solution()

    val nums = intArrayOf(2, 7, 11, 15)

    val expected = intArrayOf(0, 1)
    val result = solution.twoSum(nums, 9)

    assert(expected contentEquals result)
}

File: leetcode/0001_two_sum.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> seen;
        for (int i = 0; i < nums.size(); i++) {
            auto it = seen.find(target - nums[i]);
            if (it != seen.end()) {
                return {it->second, i};
            }

            seen[nums[i]] = i;
        }
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {2, 7, 11, 15};

    vector<int> expected = {0, 1};
    vector<int> result = solution.twoSum(nums, 9);

    assert(expected == result);
}
